[Concurrency] Suggest replacing 'async' with 'await' at call site

etcwilde opened the pull request:
It's pretty easy to typo-replace 'await' with 'async' and get a
confusing error about 'async' not being in scope. This patch updates the
diagnostic to suggest replacing 'async' with 'await' at the call-sites
so that users aren't left scratching their heads.

Fixes: rdar://72968205

etcwilde committed:
Suggest replacing 'async' with 'await' at call site

It's pretty easy to typo-replace 'await' with 'async' and get a
confusing error about 'async' not being in scope. This patch updates the
diagnostic to suggest replacing 'async' with 'await' at the call-sites
so that users aren't left scratching their heads.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test

varungandhi-apple commented on lib/Sema/PreCheckExpr.cpp:
```diff
@@ -519,7 +519,15 @@ Expr *TypeChecker::resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE,
         typo->addFixits(diag);
       } else {
-        emitBasicError();
+        if (Name.isSimpleName("async")) {
+          auto diag = Context.Diags.diagnose(Loc,
+              diag::cannot_find_in_scope_corrected, Name,
+              /*isOperator=*/false, "await");
+          diag.highlight(UDRE->getSourceRange());
+          diag.fixItReplace(Loc, "await");
```
I think you should be able to chain the method calls (in case you'd like to avoid creating a binding for `diag`).

etcwilde replied:
This is what's done immediately above. If you want it chained, I'm not terribly partial so I can make that change, but I do want to stay consistent.

varungandhi-apple replied:
Ah okay. Yeah, it was just FYI (in case you missed it). Creating a binding is fine by me. 👍🏼

varungandhi-apple commented on test/Concurrency/await_typo_correction.swift:
```diff
@@ -0,0 +1,31 @@
+
+@main struct MyProgram {
+  static func main() async throws {
+    // expected-error@+7 {{cannot find 'async' in scope; did you mean 'await'?}}{{9-14=await}}
+    // expected-error@+6 {{consecutive statements on a line must be separated by ';'}}
+    // expected-error@+5 {{call is 'async' but is not marked with 'await'}}
+    // expected-error@+4 {{call can throw but is not marked with 'try'}}
+    // expected-note@+3 {{did you mean to use 'try'?}}
+    // expected-note@+2 {{did you mean to handle error as optional value?}}
+    // expected-note@+1 {{did you mean to disable error propagation?}}
+    try async asyncFunc()
```
Couple of things:

1. IMO `cannot find 'async' in scope; did you mean 'await'?` is the wrong diagnostic message. Can we instead simplify it to (say) `found 'async' in expression, did you mean 'await'?` ? I understand that technically `async` is a contextual keyword, but we should focus on the 95% use case, where someone is very likely referring to the `async` keyword.
2. [Not strictly part of this PR but] Can we avoid a cascade of errors here? It seems really unfortunate to have 7 (!) messages all at the same spot. Despite being a compiler developer, I have no idea what the "did you mean to disable error propagation?" means...

varungandhi-apple replied:
(For the error propagation one, I see there is a fix-it to add `try!`, so it's more understandable with the fix-it, but I still think the number of messages is too high here. For example, we shouldn't be saying "did you mean to use 'try'?" if there is a `try` already present.)

etcwilde replied:
Okay, now we only emit the one error about replacing the `async` with `await.

varungandhi-apple commented on test/Concurrency/await_typo_correction.swift:
```diff
@@ -0,0 +1,31 @@
+    // expected-note@+1 {{did you mean to disable error propagation?}}
+    try async asyncFunc()
+
+    // expected-error@+4 {{consecutive statements on a line must be separated by ';'}}
+    // expected-error@+3 {{cannot find 'async' in scope; did you mean 'await'?}}{{13-18=await}}
+    // expected-warning@+2 {{result of call to 'anotherAsyncFunc()' is unused}}
+    // expected-error@+1 {{call is 'async' but is not marked with 'await'}}
+    let _ = async anotherAsyncFunc()
```
This is a slightly less worse version of the above (4 messages instead of 7), but it would be nice if we could avoid emitting `call is 'async' but is not marked with 'await'` if we already emitted the diagnostic that fixes `async -> await` earlier.

etcwilde commented:
I'm going to emit the error in the parser, fix the AST there, and then we should be able to continue compilation without emitting the confusing errors.

etcwilde committed:
Replace improper 'async' in parser

By replacing the 'async' with 'await' in the parser, we avoid the issue
of cascading errors as the compiler gets more and more confused by what
it's reading. Instead, everything mostly passes and we just emit the one
error message.

One caveat that I hadn't taken into account before was that we could
have a function called "async", in which case we don't want to replace
the "async" keyword with "await".

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
It looks like the windows bot is full. The failure was: <unknown>:0: note: IO failure on output stream: no space on device
https://ci-external.swift.org/job/swift-PR-windows/10264/console

varungandhi-apple reviewed (approved):
Mostly LGTM. It would be nice to add a test case for `async try` (adapting the code if needed to correct it to `try await`), but it's less important and would be okay as a smaller JIRA task that someone else can pick up.

etcwilde committed:
Add test case for `async try foo()`

Ensure that we emit the proper diagnostics for when the await and try
are flipped, but also await is incorrectly spelled `async`.

We're expecting two diagnostics, both of which have fix-its that do the
right thing. One will just replace the `async` with an `await`. The
other will move the text so that try comes first. Because `await` and
`async` are both have the same number of characters, both fix-its will
fix the issue correctly.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
> Mostly LGTM. It would be nice to add a test case for `async try` (adapting the code if needed to correct it to `try await`), but it's less important and would be okay as a smaller JIRA task that someone else can pick up.

Test case added. We get both diagnostics for fixing the `async` to `await` and the warning that puts the await after the `try`.
The fix-it for the warning will correct the positioning of the bad token and put an `await` in the right place, regardless of whether it was `async` or `await`.

etcwilde commented:
@swift-ci please smoke test windows platform
